pkg/datapath/garp/events.go

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package garp
+
+import (
+	"context"
+	"net/netip"
+
+	"github.com/cilium/workerpool"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+)
+
+var nodePodsCell = cell.Provide(func(lc hive.Lifecycle, c k8sClient.Clientset) (resource.Resource[*corev1.Pod], error) {
+	if !c.IsEnabled() {
+		return nil, nil
+	}
+
+	lw := utils.ListerWatcherWithFields(utils.ListerWatcherFromTyped[*corev1.PodList](
+		c.CoreV1().Pods("")), fields.OneTermEqualSelector("spec.nodeName", nodeTypes.GetName()),
+	)
+	return resource.New[*corev1.Pod](lc, lw), nil
+})
+
+// EventProcessorCell processes k8s pod events for the current node and
+// determines if a Gratuitous ARP packet needs to be sent.
+var EventProcessorCell = cell.Module(
+	"garp-event-processor",
+	"GARP event processor",
+
+	Cell,
+	nodePodsCell,
+
+	cell.Provide(newEventProcessor),
+)
+
+type EventProcessor interface {
+	Start(hive.HookContext) error
+	Stop(hive.HookContext) error
+}
+
+type eventProcessorParams struct {
+	cell.In
+
+	Logger     logrus.FieldLogger
+	Lifecycle  hive.Lifecycle
+	Pods       resource.Resource[*corev1.Pod]
+	GARPSender Sender
+	Config     Config
+}
+
+func newEventProcessor(p eventProcessorParams) EventProcessor {
+	if !p.Config.L2AnnouncePods {
+		return nil
+	}
+
+	if p.Pods == nil {
+		return nil
+	}
+
+	ep := &eventProcessor{
+		log:         p.Logger,
+		pods:        p.Pods,
+		garpSender:  p.GARPSender,
+		podIPsState: make(map[resource.Key]netip.Addr),
+	}
+
+	p.Lifecycle.Append(ep)
+
+	p.Logger.Info("initialised gratuitous arp event processor")
+
+	return ep
+}
+
+type eventProcessor struct {
+	wp *workerpool.WorkerPool
+
+	log        logrus.FieldLogger
+	pods       resource.Resource[*corev1.Pod]
+	garpSender Sender
+
+	podIPsState map[resource.Key]netip.Addr
+}
+
+func (ep *eventProcessor) Start(hive.HookContext) error {
+	ep.wp = workerpool.New(1)
```

Comment:
`pkg/hive/job` provides a nicer API for this now

Reply:
I'll use this in a followup PR, as we want to backport this.